hashes/src/sha256t.rs

```diff
@@ -1,26 +1,135 @@
 // SPDX-License-Identifier: CC0-1.0
 
-//! SHA256t implementation (tagged SHA256).
+//! SHA-256t implementation (tagged SHA-256).
 
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::vec::Vec;
 use core::cmp;
 use core::marker::PhantomData;
-use core::ops::Index;
-use core::slice::SliceIndex;
 
-use crate::{sha256, FromSliceError};
+use crate::{sha256, HashEngine};
 
-type HashEngine = sha256::HashEngine;
+/// Engine to compute tagged SHA-256 hash function.
+#[derive(Clone)]
+pub struct Engine<T>(sha256::Engine, PhantomData<T>);
 
-/// Trait representing a tag that can be used as a context for SHA256t hashes.
-pub trait Tag {
+impl<T: Tag> Default for Engine<T> {
+    fn default() -> Self { <T as Tag>::engine() }
+}
+
+impl<T: Tag> HashEngine for Engine<T> {
+    type Digest = [u8; 32];
+    type Midstate = sha256::Midstate;
+    const BLOCK_SIZE: usize = sha256::BLOCK_SIZE;
+
+    #[inline]
+    fn new() -> Self { <T as Tag>::engine() }
```

Comment:
No need to override this.

Reply:
Flagging that I read this and with the other changes to new/default I've made this module uniform.